### DISCUSSION TITLE
[General] Deploy new instances with factory contract (similar to Clones and ERC-1167: Minimal Proxy Contract)

### DISCUSSION THREAD

**Discussion by kirilaa** [zkSync-Community-Hub/zksync-developers#166] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

I am trying to deploy instances of a contract A from Factory contract B 
`B -> deploys A`

This is a common procedure ERC-1167: Minimal Proxy Contract on every EVM chain

### What is the specific issue or error you're encountering?

1. Tried to use `create2` on the Contract deployer contract using the bytecodehash of contract A manually on the blockexplorer. The is successful
2. The issue is when I proceed with the same procedure, but triggering through a function from the factory contract B. 
`reason="execution reverted: Deployment failed", method="estimateGas"`

- The `salt` used is different then the one used in step 1
- the `speedContractHash` is the same as the bytecodehash of contract A in step 1

```
import "@matterlabs/zksync-contracts/l2/system-contracts/Constants.sol";
import "@matterlabs/zksync-contracts/l2/system-contracts/libraries/SystemContractsCaller.sol";


function createJustProxyMarket(
        bytes32 salt
        ) external returns(address newContract){
        (bool success, bytes memory returnData) = SystemContractsCaller
            .systemCallWithReturndata(
                uint32(gasleft()),
                contractDeployer,
                uint128(0),
                abi.encodeCall(
                    DEPLOYER_SYSTEM_CONTRACT.create2Account,
                    (
                        salt,
                        speedContractHash,
                        abi.encode(msg.sender),
                        IContractDeployer.AccountAbstractionVersion.Version1
                    )
                )
            );
        require(success, "Deployment failed");

        (newContract) = abi.decode(returnData, (address));

    }
 ```

### Can you share the error messages or logs you're receiving, if any?

`reason="execution reverted: Deployment failed", method="estimateGas"`

### Have you made any recent changes to the contract before encountering this issue?

No

### Are there any external libraries or contracts that your contract interacts with?

No

### Can you provide the relevant portions of your contract code where the issue is occurring?

```
import "@matterlabs/zksync-contracts/l2/system-contracts/Constants.sol";
import "@matterlabs/zksync-contracts/l2/system-contracts/libraries/SystemContractsCaller.sol";


function createJustProxyMarket(
        bytes32 salt
        ) external returns(address newContract){
        (bool success, bytes memory returnData) = SystemContractsCaller
            .systemCallWithReturndata(
                uint32(gasleft()),
                contractDeployer,
                uint128(0),
                abi.encodeCall(
                    DEPLOYER_SYSTEM_CONTRACT.create2Account,
                    (
                        salt,
                        speedContractHash,
                        abi.encode(msg.sender),
                        IContractDeployer.AccountAbstractionVersion.Version1
                    )
                )
            );
        require(success, "Deployment failed");

        (newContract) = abi.decode(returnData, (address));

    }
 ```

### Have you tried to isolate the problem, and if so, what were the results?

I have a success only through manual trigger of the create2 function. 
The Discord bot suggested this solution, which does not solve the issue

![image](https://github.com/zkSync-Community-Hub/zkync-developers/assets/15322348/49c1a229-a6fd-4b72-a4f2-e7ae7863b3ea)

### What steps have you already taken to try to resolve the issue?

I have been following the bot suggestions

**Comment by MexicanAce** [comment]
Hi @kirilaa 👋 

Do you have a link to the source code? With just the snippet you've provided, it's difficult to diagnose the exact reason this is occurring.

I'm interested in why you're passing `contractDeployer` into `SystemContractsCaller.systemCallWithReturndata(...)`, could you try  `address(DEPLOYER_SYSTEM_CONTRACT)` instead?

[Check out the docs here for an example of a factory in zkSync.](https://era.zksync.io/docs/dev/tutorials/custom-aa-tutorial.html#the-factory)

**Reply by kirilaa** [reply]
Hi @MexicanAce, 
Thanks for the answer. 

The code snippet is here:
https://github.com/thales-markets/zksync-contracts/blob/main/contracts/SpeedMarkets/SpeedMarketsAMM.sol#L318

The factory contract is already deployed:
https://goerli.explorer.zksync.io/address/0x69316858c6C3c32e992C9651a6139c4814Faf82f#contract

The mastercopy contract is already deployed through the transaction:
https://goerli.explorer.zksync.io/tx/0x98fe277dc06ba129a3602b7e8f357614c04a476c73b8a6560d30d7c07dfcfe40#overview

where the `bytecodeHash` is taken and set on the factory contract

Then when `createJustProxyMarket` is called on the factory contract, the same error is thrown. 
**Note:** Several tries have been perfomed to use different variable than the `zeroBytes` (for example `abi.encode("")`) without any success.

```
function createJustProxyMarket(bytes32 salt) external returns (address newContract) {
        // _createNewProxyMarket(salt);
        bytes memory zeroBytes;
        (bool success, bytes memory returnData) = SystemContractsCaller.systemCallWithReturndata(
            uint32(gasleft()),
            address(DEPLOYER_SYSTEM_CONTRACT),
            uint128(0),
            abi.encodeCall(
                DEPLOYER_SYSTEM_CONTRACT.create2Account,
                (salt, speedContractHash, zeroBytes, IContractDeployer.AccountAbstractionVersion.Version1)
            )
        );
        require(success, "Deployment failed");

        (newContract) = abi.decode(returnData, (address));
    }
  ```

**Comment by idea404** [comment]
Hi @kirilaa,

I noticed on the transactions you shared you set values for bytecodes post-deployment. TLDR is this pattern is currently unsupported by the zkEVM. Below I give more background on what's happening here. 

**EIP 1167 and Dynamic Bytecode Declaration**

The EIP 1167 is written directly in EVM bytecode, which is quite different from the bytecode that zkEVM operates on. As a result, it's currently not feasible to use EIP 1167 on zkSync's Era. To support this pattern, we'd need to craft an entirely new "minimal" bytecode sequence. The benefits of using this minimal proxy pattern on Era might not be as significant as on Ethereum. Since deploying duplicated contracts on Era could be more cost-effective compared to Ethereum but this is something we might need to validate.

So you can't dynamically declare bytecode for factories right now [as mentioned here](https://discord.com/channels/1037725930715619418/1154362596800090152/1154411476266786877). 

**Additional Note on Fetching Deployed Bytecode**

It is also worth noting that fetching the bytecode for a deployed contract is also currently unsupported [as mentioned here](https://era.zksync.io/docs/reference/architecture/differences-with-ethereum.html#extcodecopy).

**Examples**

I made a repo to exemplify what happens when you dynamically declare bytecode for factories vs declaring it and having the compiler know of the bytecode [here](https://github.com/idea404/factory-example). 

To see run the examples and see the differences:

- Clone repo locally by running: `git clone https://github.com/idea404/factory-example`
- Change directories to the cloned project: `cd factory-example`
- Install dependencies by running `yarn`
  - Run the working example with: `yarn compile && yarn deploy:compiled`
  - Run the failing example with: `yarn compile && yarn deploy:declared`
  
You might notice that the path where the bytecode is defined statically runs without errors. On the other hand, running the dynamically declared bytecode path returns the `eth_estimateGas` error during deployment: 

```sh
...
error: Error: processing response error (body="{\"jsonrpc\":\"2.0\",\"error\":{\"code\":3,\"message\":\"execution reverted: Deployment failed\",\"data\":\"0x08c379a0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000114465706c6f796d656e74206661696c6564000000000000000000000000000000\"},\"id\":50}\n", error={"code":3,"data":"0x08c379a0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000114465706c6f796d656e74206661696c6564000000000000000000000000000000"}, requestBody="{\"method\":\"eth_estimateGas\",\"params\":[{\"gasPrice\":\"0xee6b280\",\"type\":\"0x0\",\"from\":\"0x36615cf349d7f6344891b1e7ca7c72883f5dc049\",\"to\":\"0x111c3e89ce80e62ee88318c2804920d4c96f92bb\",\"data\":\"0x76fb8b650000000000000000000000000000000000000000000000000000000000000000000000000000000000000000f85bab911139cee81fed5ffd47f37abbc420853c00000000000000000000000098e7c6fa66cc23e78a20090c77774e238621531c\"}],\"id\":50,\"jsonrpc\":\"2.0\"}", requestMethod="POST", url="http://127.0.0.1:8011", code=SERVER_ERROR, version=web/5.7.1)
        at Logger.makeError (/Users/dennis/Documents/Code/factory-example/node_modules/@ethersproject/logger/src.ts/index.ts:269:28)
        at Logger.throwError (/Users/dennis/Documents/Code/factory-example/node_modules/@ethersproject/logger/src.ts/index.ts:281:20)
        at /Users/dennis/Documents/Code/factory-example/node_modules/@ethersproject/web/src.ts/index.ts:341:28
        at step (/Users/dennis/Documents/Code/factory-example/node_modules/@ethersproject/web/lib/index.js:33:23)
        at Object.next (/Users/dennis/Documents/Code/factory-example/node_modules/@ethersproject/web/lib/index.js:14:53)
        at fulfilled (/Users/dennis/Documents/Code/factory-example/node_modules/@ethersproject/web/lib/index.js:5:58)
        at processTicksAndRejections (node:internal/process/task_queues:95:5) {
      reason: 'processing response error',
      code: 'SERVER_ERROR',
      body: '{"jsonrpc":"2.0","error":{"code":3,"message":"execution reverted: Deployment failed","data":"0x08c379a0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000114465706c6f796d656e74206661696c6564000000000000000000000000000000"},"id":50}\n',
      error: [Error],
      requestBody: '{"method":"eth_estimateGas","params":[{"gasPrice":"0xee6b280","type":"0x0","from":"0x36615cf349d7f6344891b1e7ca7c72883f5dc049","to":"0x111c3e89ce80e62ee88318c2804920d4c96f92bb","data":"0x76fb8b650000000000000000000000000000000000000000000000000000000000000000000000000000000000000000f85bab911139cee81fed5ffd47f37abbc420853c00000000000000000000000098e7c6fa66cc23e78a20090c77774e238621531c"}],"id":50,"jsonrpc":"2.0"}',
      requestMethod: 'POST',
      url: 'http://127.0.0.1:8011'
    }
  },
  tx: {
    data: '0x76fb8b650000000000000000000000000000000000000000000000000000000000000000000000000000000000000000f85bab911139cee81fed5ffd47f37abbc420853c00000000000000000000000098e7c6fa66cc23e78a20090c77774e238621531c',
    to: Promise { '0x111C3E89Ce80e62EE88318C2804920D4c96f92bb' },
    from: '0x36615Cf349d7F6344891B1e7CA7C72883F5dc049',
    type: 0,
    gasPrice: Promise { [BigNumber] },
    nonce: Promise { 2 },
    gasLimit: Promise { <rejected> [Circular *1] },
    chainId: Promise { 260 }
  }
  ...
  ```

**Reply by kirilaa** [reply]
Hi @idea404,
I examined your example carefully and tried to apply to my contracts and deployment scripts, but without any success. 
The deployed contract is [here](https://goerli.explorer.zksync.io/address/0xF98CB2eD5Ca17a4E268a6f6CFE18E12A68CDED66#contract).

Afterwards I decided to deploy the exact contracts from the example ([this ones](https://github.com/idea404/factory-example/tree/main/contracts)) and deployed them on goerli ([here](https://goerli.explorer.zksync.io/address/0xB3F752a6abD9c28BFBAAF9b57b464349B58A2C46#contract)).
The result is the same, getting the error `Deployment failed`

I have also tried doing it with script, which also failed with the same outcome:
```
Contract successfully verified on zkSync block explorer!
start deploying accounts
salt:  0x0000000000000000000000000000000000000000000000000000000000000000
owner1:  0x15E6804F226DBa9f88d080fd1E10084Ead942C1e
owner2:  0x260E3DcF6298c7D56368dd4954416b0440c25d88
<ref *1> Error: cannot estimate gas; transaction may fail or may require manual gas limit [ See: https://links.ethers.org/v5-errors-UNPREDICTABLE_GAS_LIMIT ] (error={"reason":"execution reverted: Deployment 
failed","code":"UNPREDICTABLE_GAS_LIMIT"
```
The deployment script that I used is [here](https://github.com/thales-markets/zksync-contracts/commit/b1f994164fc08291017c83cdfbff45a8ee0686e3#diff-9b5b32149b1e2367267f5a82ead67b3f5e5b0880e299cbc42254ecdb8f761216).

**Can you point me to a running example on goerli?**
If possible with a similar deployment script as in the previous example.

Thanks!

**Reply by idea404** [reply, marked as answer]
Hi @kirilaa, 

It's been a while but I've been caught up with other prios, but I adapted the example repo (quickly and I think it is actually ugly code but it works) and deployed a working version to testnet. So, to reproduce you will need to first clone the example repo again: https://github.com/idea404/factory-example. Then, this is the sequence of commands I ran:

1. Deploy the factory to testnet
```bash
WALLET_KEY=<YOUR-TESTNET-WALLET-PRIVATE-KEY> yarn deploy:compiled:factory --network zkSyncTestnet
yarn run v1.22.19
$ yarn hardhat deploy-zksync --script deploy-factory.ts --network zkSyncTestnet
$ /Users/dennis/Documents/Code/factory-example/node_modules/.bin/hardhat deploy-zksync --script deploy-factory.ts --network zkSyncTestnet
AA factory address: 0xDE7926Ad640005C33843B5EC178DC757e7079957
✨  Done in 8.59s.
```

2. Deploy the multisig
```bash
WALLET_KEY=<YOUR-TESTNET-WALLET-PRIVATE-KEY> yarn deploy:compiled:multisig --network zkSyncTestnet
yarn run v1.22.19
$ yarn hardhat deploy-zksync --script deploy-multisig.ts --network zkSyncTestnet
$ /Users/dennis/Documents/Code/factory-example/node_modules/.bin/hardhat deploy-zksync --script deploy-multisig.ts --network zkSyncTestnet
Multisig account deployed on address 0x613E05Cbe1b2a0FF29776dF08D4fB76CB0Ed78A0
Sending funds to multisig account
Multisig account balance is 8000000000000000
The multisig's nonce before the first tx is 0
The multisig's nonce after the first tx is 1
Multisig account balance is now 7949497000000000
✨  Done in 13.08s.
```

So both addresses of deployed contracts:

- Factory: https://zksync2-testnet.zkscan.io/address/0xDE7926Ad640005C33843B5EC178DC757e7079957/transactions
- MultiSig: https://zksync2-testnet.zkscan.io/address/0x613E05Cbe1b2a0FF29776dF08D4fB76CB0Ed78A0/transactions

**Reply by kirilaa** [reply]
Hi @idea404, 
Just tried all the code that you have sent me. 
On the second step, I still get the same error:
`error={"reason":"execution reverted: Deployment failed",`

**Reply by kirilaa** [reply]
On a clean repo, it works. However the point is to have a factory contract for creating new contracts. 
I am gonna try to explore if that is possible departing from your example.

**Comment by rietracerdanchri** [comment]
It seems like the issue might be related to the way the salt is generated or used in the createJustProxyMarket function. Ensure that the salt passed to the create2 function is consistent and correctly derived in both manual triggering and when called from the factory contract. Additionally, verify that all other parameters, such as speedContractHash and the encoded data, are correctly formatted and passed to the create2Account function. Double-checking these parameters and their consistency between manual triggering and contract function invocation should help resolve the deployment failure issue.

**Reply by kirilaa** [reply]
At the end, the architecture was refactored to avoid creation of new clones/instances.   
Solution already in production/ on-chain on ZkSync Era Mainnet. 

Minimal proxy instances seem to be a breaking point.

**Comment by marqbassmier** [comment]
Ensure that the salt value used in the create2 call within the factory contract B matches the salt value used in the successful manual trigger of the create2 function. Consistent salt values are crucial for deploying contracts deterministically.

**Comment by singletonbrittany** [comment]
It seems like the issue might be related to the salt parameter used in the create2 call. Ensure that the salt value passed from the factory contract B matches the one used successfully in the manual trigger of the create2 function. Double-checking the salt value could help resolve the deployment failure issue.
